### PR TITLE
nixos: Introduce nix.buildLocation option

### PR DESCRIPTION
**Pull request by avnik**
Allow to specify where package build will happens.
It helps big packages (like browsers) not to overflow tmpfs.

###### Motivation for this change

/tmp overflow during build big packages (like chromium and/or firefox) is very annoying, so I added option to run builds on dedicated filesystem/directory. 

/cc @cleverca22 for review, we discuss possibility of this change earlier.
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by volth** [comment]
> /tmp overflow during build big packages (like chromium and/or firefox) is very annoying

better to blame `tmpfs` and its hard limit of disk usage

`boot.tmpOnTmpfs = false` does solve the issue but at the price of performance degradation and increased disk wear.
`nix.buildLocation` effectively does the same.

It might be better to find a way to tune `tmpfs` increasing or removing its disk limits (it uses the disk after RAM limit is reached)

**Comment by avnik** [comment]
Well, I prefer (ideally) to force nix-daemon to choose  /tmp, /var/ssd, or /var/spinning-disk depending of size of sources, and predicted overall usage.

I definitely won't to degrade performance of everything, pushing /tmp to spinning disks.  I actually had this solutuion, now generalized to nixos option, which I used couple years.  We discussed it in IRC yesterday, and I proposed to make this chanhe (it why I cast @cleverca22 here).

But you are right -- is an opinionated change, and point to discuss,

**Comment by volth** [comment]
BTW, `tmpfs` can be resized on the fly (`mount -o remount,size=50G /tmp`).
nix-daemon could manage it, or create own `tmpfs` mount point for the time of build

File: nixos/modules/services/misc/nix-daemon.nix [hunk 1]
```diff
@@ -428,6 +441,12 @@ in
             LimitNOFILE = 4096;
           };
 
+        preStart = ''
+          # ignore error and continue with error message. Path can be manually created to recovery.
+          # otherwise user wouldn't be able fix path and rebuild.
+          mkdir -p ${cfg.buildLocation} || echo "nix-daemon pre-start: can not create ${cfg.buildLocation}" >&2
```
Comment (Mic92) [inline nixos/modules/services/misc/nix-daemon.nix]: systemd-tmpfiles might be a better solution here.

File: nixos/modules/services/misc/nix-daemon.nix [hunk 2]
```diff
@@ -417,7 +428,9 @@ in
           ++ optionals cfg.distributedBuilds [ pkgs.gzip ];
 
         environment = cfg.envVars
-          // { CURL_CA_BUNDLE = "/etc/ssl/certs/ca-certificates.crt"; }
+          // { CURL_CA_BUNDLE = "/etc/ssl/certs/ca-certificates.crt";
+               TMPDIR = "${cfg.buildLocation}";
```
Comment (Mic92) [inline nixos/modules/services/misc/nix-daemon.nix]:
```suggestion
               TMPDIR = cfg.buildLocation;
```

**Comment by avnik** [comment]
@volth I see, but anyway it not solve  memory vs swap usage here.  Ideally would be nice to have ability to "annotate" derivation by some labels, (which should not  affect purity/cache), but suggest nix to pick better builder (or builder options -- like spool dir/cpu cores/memory/time limits, etc)

**Comment by avnik** [comment]
@Mic92  applied both suggestions

**Comment by Mic92** [comment]
@GrahamcOfBorg test installer.simple

**Comment by Mic92** [comment]
@GrahamcOfBorg test installer.simple

**Comment by Mic92** [comment]
This has a merge conflict now unfortunately.

File: nixos/modules/services/misc/nix-daemon.nix [hunk 3]
```diff
@@ -490,6 +502,8 @@ in
         restartTriggers = [ nixConf ];
       };
 
+    systemd.tmpfiles.rules = [ "d ${cfg.buildLocation} 0775 root root -" ];
```
Comment (gravndal) [inline nixos/modules/services/misc/nix-daemon.nix:505]: This breaks write permissions for non-root on `/tmp` with the default value and `boot.tmpOnTmpfs = true;`
```
setting up tmpfiles
[/etc/tmpfiles.d/tmp.conf:11] Duplicate line for path "/tmp", ignoring.
```
```
$ getfacl /tmp
getfacl: Removing leading '/' from absolute path names
# file: tmp/
# owner: root
# group: root
user::rwx
group::rwx
other::r-x
```

**Comment by primeos** [comment]
As @gravndal already noted the `tmpfiles` change is problematic and could cause severe regressions (I didn't update yet, but I'd imaging many applications might crash and/or missbehave due to this).

See also #85552.
I think this should be fixed or reverted ASAP to reduce the impact.

**Comment by calbrecht** [comment]
Yes :) it is far worse. This affects every reboot. Sway segfaults for me when it cannot write to /tmp.

**Comment by primeos** [comment]
I reverted this change in 0e4417f1185e2093eb4f084e144f49a66f701db9 for now so we can properly resolve this without any pressure.

**Comment by Mic92** [comment]
@primeos thanks!

**Comment by avnik** [comment]
@Mic92 @primeos should I re-do other PR with change to preStart?

**Comment by Mic92** [comment]
yes.

**Pull request merged**